relay/nakama/custom_auth.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+
+	"github.com/heroiclabs/nakama-common/api"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/rotisserie/eris"
+	"google.golang.org/grpc/codes"
+
+	"pkg.world.dev/world-engine/relay/nakama/siwe"
+	"pkg.world.dev/world-engine/relay/nakama/utils"
+)
+
+const (
+	signInWithEthereumType = "siwe"
+)
+
+var (
+	ErrBadCustomAuthType = errors.New("bad custom auth type")
+)
+
+func initCustomAuthentication(initializer runtime.Initializer) error {
+	if err := initializer.RegisterBeforeAuthenticateCustom(handleCustomAuthentication); err != nil {
+		return eris.Wrap(err, "failed to init siwe authentication")
+	}
+	return nil
+}
+
+func handleCustomAuthentication(
+	ctx context.Context,
+	logger runtime.Logger,
+	_ *sql.DB,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest) (*api.AuthenticateCustomRequest, error) {
+	authType := in.GetAccount().GetVars()["type"]
+	// In the future, other authentication methods can be added here (e.g. Twitter)
+	if authType == signInWithEthereumType {
+		return handleSIWE(ctx, logger, nk, in)
+	}
+	return nil, ErrBadCustomAuthType
+}
+
+func handleSIWE(
+	ctx context.Context,
+	logger runtime.Logger,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest,
+) (
+	*api.AuthenticateCustomRequest, error,
+) {
+	signerAddress := in.GetAccount().GetId()
+	signature := in.GetAccount().GetVars()["signature"]
+	message := in.GetAccount().GetVars()["message"]
+
+	if signerAddress == "" {
+		_, err := utils.LogErrorWithMessageAndCode(
+			logger, siwe.ErrMissingSignerAddress, codes.InvalidArgument, "id field must be set")
+		return nil, err
+	}
+	if signature == "" && message != "" {
+		_, err := utils.LogErrorWithMessageAndCode(
+			logger, siwe.ErrMissingSignature, codes.InvalidArgument, "missing field")
+		return nil, err
+	}
+	if signature != "" && message == "" {
+		_, err := utils.LogErrorWithMessageAndCode(
+			logger, siwe.ErrMissingMessage, codes.InvalidArgument, "missing field")
+		return nil, err
+	}
+
+	if signature != "" && message != "" {
+		// The user has provided a signature and a message. Attempt to authenticate the user.
+		if err := siwe.ValidateSignature(ctx, nk, signerAddress, message, signature); err != nil {
+			_, err = utils.LogErrorWithMessageAndCode(
+				logger, siwe.ErrMissingMessage, codes.Unauthenticated, "authentication failed")
+			return nil, err
+		}
+		return in, nil
+	}
+
+	// The signature and message is empty. Generate a new SIWE message for the user.
```

Comment:
This might be a personal preference, but I typically think error states like this should be handled first and terminates the function earlier before going with the happy path

Reply:
done. Error/siwe generation has been moved up. The last operation in this function is now the successful authentication of a user.